File: apps/burner-windows/src/PairSession.cs
using System;
using System.IO;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace Flagship.Burner;

/// <summary>
/// One phone-pairing milestone, parsed from a `FLAGSHIP_PAIR <json>` line the
/// `flagship-burn pair --emit-events` subprocess prints. Mirrors the shared
/// PairEvent union in packages/flagship-burner/src/pair.ts.
/// </summary>
public sealed record PairEvent
{
    public required string Event { get; init; }
    // ready
    public string? HumanCode { get; init; }
    public string? QrTerminal { get; init; }
    public string? Payload { get; init; }
    public bool DebugRequested { get; init; }
    // phone-connected
    public string? Sas { get; init; }
    // delivered / done
    public string? ServerDomain { get; init; }
    // done
    public string? RecipePath { get; init; }
    public bool DebugGranted { get; init; }
    // error
    public string? Message { get; init; }
}

/// <summary>
/// Pure parser for the subprocess's structured stdout. Kept separate from the
/// process plumbing so it's unit-testable. A line that isn't a well-formed
/// `FLAGSHIP_PAIR <json>` returns null (ordinary human log line).
/// </summary>
public static class PairEventParser
{
    public const string Prefix = "FLAGSHIP_PAIR ";

    public static PairEvent? TryParse(string line)
    {
        if (line is null) return null;
        var trimmed = line.TrimStart();
        if (!trimmed.StartsWith(Prefix, StringComparison.Ordinal)) return null;
        var json = trimmed[Prefix.Length..];
        try
        {
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object) return null;
            if (!root.TryGetProperty("event", out var ev) || ev.ValueKind != JsonValueKind.String)
                return null;
            string? Str(string name) =>
                root.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;
            bool Bool(string name) =>
                root.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.True;
            return new PairEvent
            {
                Event = ev.GetString()!,
                HumanCode = Str("humanCode"),
                QrTerminal = Str("qrTerminal"),
                Payload = Str("payload"),
                DebugRequested = Bool("debugRequested"),
                Sas = Str("sas"),
                ServerDomain = Str("serverDomain"),
                RecipePath = Str("recipePath"),
                DebugGranted = Bool("debugGranted"),
                Message = Str("message"),
            };
        }
        catch (JsonException)
        {
            return null;
        }
    }
}

/// <summary>
/// Drives `flagship-burn pair --out &lt;tmp&gt; --emit-events` and surfaces its
/// milestones as .NET events. The heavy lifting (the relay handshake, SAS
/// derivation, recipe decrypt + signature verify) is the SHARED TypeScript
/// implementation — this is thin glue, exactly like the remaster/write path.
/// On the terminal `done` event the recipe JSON has been written to the temp
/// path; the caller loads + verifies it locally (RecipeLoader) and proceeds to
/// the destination chooser.
/// </summary>
public sealed class PairSession
{
    private readonly bool _debug;
    private CliRunner? _runner;
    private readonly CancellationTokenSource _cts = new();

    /// <summary>Raised on the UI-marshalling caller's thread is NOT guaranteed —
    /// subscribers must dispatch to the UI themselves (the Wizard does).</summary>
    public event Action<PairEvent>? OnEvent;
    public event Action<LogLine>? OnLog;

    public string RecipeOutPath { get; }

    public PairSession(bool debug, string? outPath = null)
    {
        _debug = debug;
        RecipeOutPath = outPath ?? Path.Combine(Path.GetTempPath(),
            $"flagship-pair-{Guid.NewGuid():N}.json");
    }

    /// <summary>
    /// Spawn the pairing subprocess. Completes when it exits (after `done` or an
    /// error). Throws only if the CLI can't be located/spawned.
    /// </summary>
    public async Task RunAsync()
    {
        var resolved = CliLocator.Locate();
        var args = CliArgs.Pair(resolved.EntryPath, RecipeOutPath, _debug);
        _runner = new CliRunner(resolved.NodePath, args);
        await _runner.RunAsync(line =>
        {
            var ev = line.Stream == LogStream.Stdout ? PairEventParser.TryParse(line.Text) : null;
            if (ev != null) OnEvent?.Invoke(ev);
            else OnLog?.Invoke(line);
        }, _cts.Token);
    }

    public void Cancel()
    {
        try { _cts.Cancel(); } catch { }
        _runner?.Cancel();
    }
}

File: apps/builder-linux/pair_session.py
"""Phone pairing: drive `flagship-build pair --emit-events` and surface its
milestones as callbacks.

The heavy lifting (the relay handshake, SAS derivation, recipe decrypt +
signature verify) is the SHARED TypeScript implementation
(packages/flagship-builder/src/pair.ts) — this is thin glue, exactly like the
remaster/write path. Mirrors apps/builder-windows/src/PairSession.cs.

The parser is pure and unit-tested; the event shape must track the PairEvent
union in pair.ts. NOTE the field asymmetry in the TS union: the mid-session
`debug-result` event carries `granted`, while the terminal `done` event carries
`debugGranted` — both are parsed (the Windows parser misses `granted`; see the
session report).
"""
from __future__ import annotations

import json
import os
import tempfile
import threading
import uuid
from dataclasses import dataclass
from typing import Callable, Optional

from cli_runner import CLIRunner, LogLine, Resolved, args_pair, locate

PAIR_EVENT_PREFIX = "FLAGSHIP_PAIR "


@dataclass(frozen=True)
class PairEvent:
    """One phone-pairing milestone, parsed from a `FLAGSHIP_PAIR <json>` line
    the subprocess prints."""

    event: str
    # ready
    human_code: Optional[str] = None
    qr_terminal: Optional[str] = None
    payload: Optional[str] = None
    debug_requested: bool = False
    # phone-connected
    sas: Optional[str] = None
    # delivered / done
    server_domain: Optional[str] = None
    # debug-result
    granted: bool = False
    # done
    recipe_path: Optional[str] = None
    debug_granted: bool = False
    # error
    message: Optional[str] = None


def parse_pair_event(line: str) -> Optional[PairEvent]:
    """Pure parser for the subprocess's structured stdout. A line that isn't a
    well-formed `FLAGSHIP_PAIR <json>` returns None (ordinary human log
    line)."""
    if line is None:
        return None
    trimmed = line.lstrip()
    if not trimmed.startswith(PAIR_EVENT_PREFIX):
        return None
    try:
        obj = json.loads(trimmed[len(PAIR_EVENT_PREFIX):])
    except (json.JSONDecodeError, ValueError):
        return None
    if not isinstance(obj, dict):
        return None
    event = obj.get("event")
    if not isinstance(event, str):
        return None

    def s(name: str) -> Optional[str]:
        v = obj.get(name)
        return v if isinstance(v, str) else None

    def b(name: str) -> bool:
        return obj.get(name) is True

    return PairEvent(
        event=event,
        human_code=s("humanCode"),
        qr_terminal=s("qrTerminal"),
        payload=s("payload"),
        debug_requested=b("debugRequested"),
        sas=s("sas"),
        server_domain=s("serverDomain"),
        granted=b("granted"),
        recipe_path=s("recipePath"),
        debug_granted=b("debugGranted"),
        message=s("message"),
    )


class PairSession:
    """Spawn the pairing subprocess and route its output: milestone lines to
    `on_event`, everything else to `on_log`. On the terminal `done` event the
    recipe JSON has been written to `recipe_out_path`; the caller loads it
    (identical to a dropped-in recipe file) and proceeds to the destination
    chooser."""

    def __init__(
        self,
        debug: bool,
        out_path: Optional[str] = None,
        locate_fn: Callable[[], Resolved] = locate,
    ) -> None:
        self._debug = debug
        self._locate_fn = locate_fn
        self._runner: Optional[CLIRunner] = None
        self._cancelled = threading.Event()
        self.recipe_out_path = out_path or os.path.join(
            tempfile.gettempdir(), f"flagship-pair-{uuid.uuid4().hex}.json"
        )

    def run(
        self,
        on_event: Callable[[PairEvent], None],
        on_log: Callable[[LogLine], None],
    ) -> int:
        """Blocking: spawn, stream until exit (after `done` or an error).
        Callbacks fire on the runner's reader threads — callers marshal to
        their UI loop themselves. Raises only if the CLI can't be
        located/spawned."""
        resolved = self._locate_fn()
        args = args_pair(resolved.entry_path, self.recipe_out_path, self._debug)
        runner = CLIRunner(node_path=resolved.node_path, arguments=args)
        self._runner = runner

        def route(line: LogLine) -> None:
            ev = parse_pair_event(line.text) if line.stream == "stdout" else None
            if ev is not None:
                on_event(ev)
            else:
                on_log(line)

        try:
            runner.start(on_line=route)
            return runner.wait()
        finally:
            self._runner = None

    def cancel(self) -> None:
        self._cancelled.set()
        runner = self._runner
        if runner is not None:
            try:
                runner.terminate()
            except Exception:
                pass
